### PR TITLE
Clarify OnAckReceived pseudocode

### PR DESCRIPTION
**Pull request by Ralith**
I ran into a few points of confusion in the existing code:
- The time to update and scope of `largest_acked_packet` is not explicit
- `ProcessECN` references elements of `sent_packets` after they are removed by `OnPacketAcked`
- `DetectLostPackets` is passed an undefined variable

The fixes are trivial, except for the last: it isn't clear to me whether `DetectLostPackets` was intended to be called with `ack.largest_acked` or `largest_acked_packet`. These only differ when ACKs are reordered, and because the prior invocation of `DetectLostPackets` with a higher `largest_acked` would have already deemed lost all packets that the later invocation might deem lost, I don't think the behavior differs even then. Hence, removing the argument and consistently referencing `largest_acked_packet` seems to be the simpler option.

### CODE REVIEW

File: draft-ietf-quic-recovery.md [hunk 1]
```diff
@@ -720,9 +720,14 @@ Pseudocode for OnAckReceived and UpdateRtt follow:
     // ack-eliciting, update the RTT.
     if (sent_packets[ack.largest_acked] &&
         sent_packets[ack.largest_acked].ack_eliciting):
+      largest_acked_packet = ack.largest_acked
```
Comment (martinthomson) [inline draft-ietf-quic-recovery.md]: This doesn't work because the largest_acked in the ACK might be newly acknowledged, but not the largest ever acknowledged.

```suggestion
      if ack.largest_acked > largest_acked_packet:
         largest_acked_packet = ack.largest_acked
```

File: draft-ietf-quic-recovery.md [hunk 2]
```diff
@@ -720,9 +720,14 @@ Pseudocode for OnAckReceived and UpdateRtt follow:
...
       latest_rtt = now - sent_packets[ack.largest_acked].time
       UpdateRtt(latest_rtt, ack.ack_delay)
 
+    // Process ECN information if present.
+    if (ACK frame contains ECN information):
+       ProcessECN(ack)
```
Comment (martinthomson) [inline draft-ietf-quic-recovery.md]: This is a separate change.  I think that you should make this a separate PR.

**Comment by Ralith** [comment]
Split the `ProcessECN` tweak into https://github.com/quicwg/base-drafts/pull/2142, corrected the `largest_acked_packet` update logic, and added initialization for `largest_acked_packet`.

File: draft-ietf-quic-recovery.md [hunk 3]
```diff
@@ -634,7 +634,7 @@ largest_sent_packet:
 : The packet number of the most recently sent packet.
 
 largest_acked_packet:
-: The largest packet number acknowledged in an ACK frame.
+: The largest packet number acknowledged by any ACK frame.
```
Comment (janaiyengar) [inline draft-ietf-quic-recovery.md]:
```suggestion
: The largest packet number acknowledged in the connection so far.
```

Comment (martinthomson) [reply draft-ietf-quic-recovery.md]: Isn't this in the packet number space?

Comment (martinthomson) [reply draft-ietf-quic-recovery.md]: I think that you mean "packet number space" rather than connection.

File: draft-ietf-quic-recovery.md [hunk 4]
```diff
@@ -716,6 +717,9 @@ Pseudocode for OnAckReceived and UpdateRtt follow:
 
 ~~~
   OnAckReceived(ack):
+    largest_acked_packet = max(largest_acked_packet,
```
Comment (ianswett) [inline draft-ietf-quic-recovery.md]: I think this accurately describes the text we have, but I think we should add some text that you MAY use the largest acked from the most recent ACK, which could reduce spurious retransmits when one packet arrives far out of order?

Comment (ianswett) [reply draft-ietf-quic-recovery.md]: To clarify, I'm ok doing this in a follow up PR.

Comment (Ralith) [reply draft-ietf-quic-recovery.md]: > you MAY use the largest acked from the most recent ACK, which could reduce spurious retransmits when one packet arrives far out of order

Any reason not to make this the sole recommendation? I had some trouble coming up with a case where that might occur, which is why I didn't take that route originally, but I think I see it now: when a reordered ACK newly acknowledges a packet such that the newly acknowledged packet is > ack.largest_acked - kPacketThreshold, but <= largest_acked_packet - kPacketThreshold.

Comment (Ralith) [reply draft-ietf-quic-recovery.md]: Er, wait, that's nonsense. Still having trouble seeing the case.

Comment (ianswett) [reply draft-ietf-quic-recovery.md]: It's the case when one packet is far out of order, and then subsequent packets arrive in order.  ie: 100 is received, then 1, 2, 3, 4, 5, etc.  I don't fully understand the cause, but I've seen small UDP packets jump ahead of large ones in a number of traces.

But I suspect Jana and I will merge this as is and we can add follow-up text about that issue if we think it's valuable.
